pantab/_reader.py

```diff
@@ -53,6 +53,9 @@ def _read_query_result(
     for k, v in dtypes.items():
         if v == "date":
             dtypes[k] = "datetime64[ns]"
+        elif v == 'datetime64[ns, UTC]':
```

Comment:
Yup, looks to me that when I revert everything to TOT, I get a bunch of failures all attributed to the type error I listed in the bug. Here's my test report 
```bash
$ pytest pantab

FAILED pantab/tests/test_roundtrip.py::test_multiple_tables[table_name4-a] - TypeError: Cannot use .astype to convert from timezone-naive dtype to timezone-aware dtype. Use obj.tz_localize instead or series.dt.tz_localize instead
...
============================================================================================================================ 28 failed, 39 passed, 2 skipped, 4 warnings in 47.76s =============================================================================================================================
```
I followed contribute.md to set up my local env as such. 
```bash
$ conda env create -f environment.yml
$ conda activate pantab-dev
$ python setup.py build_ext --inplace
$ gcc --version
gcc (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE
$ conda list | grep pandas
pandas                    2.0.3           py311h320fe9a_1    conda-forge
```
I'll get to patching it, will mark for review when complete

Reply:
After implementing what you suggested on _readers.py, I fixed the type errors and am now running into the following:
```bash
Name: datetime64_utc, dtype: datetime64[ns]
0   2018-01-01 00:00:00+00:00
1   2019-01-01 00:00:00+00:00
2                         NaT
Name: datetime64_utc, dtype: datetime64[ns, UTC]
0   1970-01-02 15:03:06.364787+00:00
1   1970-01-02 15:03:06.364787+00:00
2                                NaT
```
We can see here that utc is starting nonsensically at beginning of epoch (or close to).
I'm now pretty sure the problem is actually with writing utc timestamps. Here's why I think that:

  - I set a breakpoint in [test_roundtrips](https://github.com/abhinavDhulipala/pantab/blob/main/pantab/tests/test_roundtrip.py#L24) after the initial writes. This let me pause the program and der ref the tmp hyper
  - I then wrote a small script use the standard tableau api to inspect the tmp table and observed the following:
 ```python
from tableauhyperapi import HyperProcess, Connection, Telemetry

with HyperProcess(Telemetry.DO_NOT_SEND_USAGE_DATA_TO_TABLEAU) as hyper:
    with Connection(hyper.endpoint, database="/tmp/pytest-of-<myhost>/pytest-8/test_basic_table_w_0/test.hyper") as connection:
        resp = connection.execute_list_query(
            query=f"""select datetime64_utc from {'"public"."table"'}"""
        )
        print(resp)
>>> [[Timestamp(1970, 1, 2, 14, 59, 40, 711575, UTC)], [Timestamp(1970, 1, 2, 14, 59, 40, 711574, UTC)], [None]]
```

Here we can see, even before reading that the writers are adding nonsensical UTC timestamps. I tried for a while to debug the writer, but that's the progress I made today on this. Do you have any ideas on where in the writer this could be going wrong?

Reply:
Huh that's strange. Can you try locally to just write a Hyper file with a UTC-aware timestamp, open it in Tableau and see what comes through? Seems like either the reader or the writer has gotten out of whack with pandas 2.0; knowing which would help us target the issue

Reply:
Are you referring to do this through pantab or natively with the Tableau hyper api?

Reply:
Essentially I am asking to try:

```python
df = pd.DataFrame([[pd.Timestamp("today", tz="UTC")]], columns=["a"])
pantab.frame_to_hyper(df, "example.hyper", table="example")
```

Then open the Hyper file in Tableau and see if it displays today's date or something closer to the Unix epoch.

If it shows today's date, then likely our writer is correct and the reader is wrong. If it shows closer to the epoch, then likely the writer is wrong (though maybe still both)

Reply:
I put the process as a failing test to isolate the write error in the latest commit. I'm pretty confident it's with the write api

Reply:
Ah ok thanks. I am on the road for a few days so won't have time to review, but hoping the issue is in the Python layer. 

If not can look at the C extension. If you've never done Python extensions before you might get some value out of this blog post on how to debug

https://willayd.com/fundamental-python-debugging-part-2-python-extensions.html

Reply:
Hmm if you are referring to the print messages in the current PR those wouldn't prove either way if it is the writer or the reader, since they are in the roundtrip test and happen after both are done.

Do you have Tableau Desktop installed on your computer? Were you able to try the steps I mentioned above and look at the file manually in that tool?

Reply:
Doesn't this [test](https://github.com/innobi/pantab/pull/187/files#diff-bd26cb0f56667586383dade9aa11e5009f4e5045c2d5ab67fb6f8b4ca9cf4432R168-R185) prove that the write api has a bug. I'm using the native library to read. the hyper, which is the same thing that'd be done by tableau itself. Sorry, I'm trying to patch this for another team. I don't directly have access to the tableau gui. But using the Hyper API as a ground truth seems fine right?

Reply:
It's also interesting to note that the test I added passes with use_parquet set to true

Reply:
Ah sorry I was looking at the print statements not the test you added - pretty clever! And yes I agree

As far as parquet goes, that isn't necessarily that surprising. parquet read/write is a newer feature of the Tableau Hyper API and will follow different code paths, at the cost of requiring users to create the parquet file

Reply:
OK so yea seems to be a problem in `writer.c`; if you drop a debugger into that function around L196 you will see that there is quite a stark difference in the `val` variable when using pandas 2.0 and UTC timestamps versus naive timestamps.

Haven't had time to look much further but hope that helps

Reply:
Appreciate that thanks, I'll look into it

Reply:
This issue might have actually started back in the 1.5 series. With this input:

```python
>>> import pandas as pd
>>> df = pd.DataFrame({"utc_time": [pd.Timestamp("today", tz="UTC")]})
```

pandas 1.4x yields:

```python
>>> df._mgr.column_arrays[0]
array(['2023-07-31T12:14:56.674835000'], dtype='datetime64[ns]')
```

Whereas 1.5 yields:

```python
>>> df._mgr.column_arrays[0]
array([Timestamp('2023-07-28 21:49:30.504629+0000', tz='UTC')],
      dtype=object)
```

In 1.4 pandas would silently truncate the timezone, and we would leverage that since the only timezone that Tableau stores things in is UTC, so we would assume timezone naive == UTC.

Looks like in 1.5 pandas wanted to be more careful about how that was handled. 

Though not ideal, I think for now we can just remove the timezone in the Python layer before calling the writer extension for any UTC timestamps

Reply:
I was looking at the writer implementation and no where does it ever pull an offset from the numpy datetime struct. I was wondering how it ever worked to be honest for any datetime with none-null(`None`) TZinfo. Also in `writer.c` I see that utc and naive are parsed as the same [info](https://github.com/abhinavDhulipala/pantab/blob/main/pantab/src/writer.c#L339-L360) with no tzinfo parsed. What am I missing here?